_posts/2016-08-02-testing-spark-streaming-unit-testing.md

```diff
@@ -0,0 +1,126 @@
+
+Controlling the lifecycle of Spark can be cumbersome and tedious. Fortunately, [Spark Testing Base](https://github.com/holdenk/spark-testing-base) project offers us Scala Traits that handle those low level details for us. Streaming has an extra bit of complexity as we need to produce data for ingestion in a timely way. At the same time, Spark internal clock needs to tick in a controlled way if we want to test timed operations as sliding windows.
+
+Let's see how to test the archetypical example of WordCount:
+
+```scala
+def count(lines: DStream[String]): DStream[(String, Int)] =
+ lines.flatMap(_.split(" "))
+   .map(word => (word, 1))
+   .reduceByKey(_ + _)
+```
+
+As you can see, this is pure function, with no side effects or access to external state. We can reason about it by having a look to the signature of the function. [DStream](https://spark.apache.org/docs/0.7.2/api/streaming/spark/streaming/DStream.html) is the basic abstraction in Spark Streaming and Spark Testing Base will help us to deal with it.
+
+```scala
+class WordCountSpec extends StreamingSuiteBase {
+
+ test("count words") {
+  val input = List(List("the word the"))
+  val expected = List(List(("the", 2), ("word", 1)))
+  testOperation[String, (String, Int)](input, count _ , expected, ordered = false)
+ }
+}
+```
+
+You don't need to work directly with DStream abstraction. The input will be a sequence of input collections, and every collection will be consumed with a tick of Spark Streaming internal clock. You can find more examples about what you can do with this library [here](https://github.com/holdenk/spark-testing-base/wiki/StreamingSuiteBase).
+
+## Joining Streaming and Batch Processing
+
+One classical scenario in Stream Processing is joining a stream with a database in order to enrich, filter or transform the events contained on the stream. Thanks to [Spark 2.0](https://spark.apache.org/releases/spark-release-2-0-0.html) and [Structured Streaming](https://jaceklaskowski.gitbooks.io/mastering-apache-spark/content/spark-sql-structured-streaming.html), Streaming and Batch are aligned, and somehow hidden, in a layer of abstraction.
+
+As Spark 2.0 has been recently released, let's focus in an example of the old API:
+
+```scala
+def countWithSpecialWords(lines: DStream[String], specialWords: RDD[String]):
+ DStream[(String, Int)] = {
+ val words = lines.flatMap(_.split(" "))
+ val bonusWords = words.transform(_.intersection(specialWords))
+
+ words.union(bonusWords)
+    .map(word => (word, 1))
+    .reduceByKey(_ + _)
+}
+```
+
+This is a convoluted example, but serves as a demonstration. Our system keeps a list of special words in an external database. We want to count twice words in the stream that are contained in that special words bag. It's important to note that our function is free of any concern about how to retrieve those special words. That's done outside of the function and that gives us the chance of unit test the logic.
```

Comment:
That's done outside of the function and that gives us the chance `to` unit test the logic.